src/Raven.Server/Documents/AI/Embeddings/QueryEmbeddingsBatchingWorker.cs

```diff
@@ -0,0 +1,283 @@
+
+        private readonly ConcurrentQueue<QueryEmbeddingsRequest> _requestQueue = new();
+        private readonly AutoResetEvent _workAvailable = new(false);
+
+        private readonly Task[] _workerTasks;
+        
+        private readonly SemaphoreSlim _queueReaderLock = new(1, 1);
+
+        public QueryEmbeddingsBatchingWorker(string databaseName,
+            AiConfiguration configuration,
+            ITextEmbeddingGenerationService service,
+            AiConnectionStringIdentifier connectionStringIdentifier,
+            RavenLogger logger,
+            CancellationToken shutdown) : base(nameof(QueryEmbeddingsBatchingWorker), logger, shutdown)
+        {
+            _databaseName = databaseName;
+            _configuration = configuration;
+            _service = service;
+            _connectionStringIdentifier = connectionStringIdentifier;
+
+            int workerCount = configuration.QueryEmbeddingsMaxConcurrentBatches;
+            _workerTasks = new Task[workerCount];
+
+            if (logger.IsInfoEnabled)
+                logger.Info($"Initializing {nameof(QueryEmbeddingsBatchingWorker)} for connection '{_connectionStringIdentifier.Value}' in database '{databaseName}' with {workerCount} workers");
+        }
+
+        public Task<ReadOnlyMemory<float>[]> EnqueueRequestAsync(IList<string> values, CancellationToken cancellationToken)
+        {
+            CancellationToken.ThrowIfCancellationRequested();
+            
+            var request = new QueryEmbeddingsRequest(values, cancellationToken);
+
+            _requestQueue.Enqueue(request);
+
+            _workAvailable.Set();
+
+            return request.TaskCompletionSource.Task;
+        }
+
+        private async Task WorkerLoopAsync()
+        {
+            while (CancellationToken.IsCancellationRequested == false)
+            {
+                try
+                {
+                    WaitHandle.WaitAny([_workAvailable, CancellationToken.WaitHandle]);
+
+                    CancellationToken.ThrowIfCancellationRequested();
+
+                    if (await _queueReaderLock.WaitAsync(0, CancellationToken))
+                    {
+                        List<QueryEmbeddingsRequest> requests;
+                        var totalNumberOfValues = 0;
+
+                        var mightBeMore = false;
+
+                        try
+                        {
+                            requests = new List<QueryEmbeddingsRequest>();
+
+                            while (_requestQueue.TryDequeue(out var request)) // todo: take into account model's token limit
+                            {
+                                if (request.TaskCompletionSource.Task.IsCanceled)
+                                    continue;
+
+                                requests.Add(request);
+
+                                totalNumberOfValues += request.Values.Count;
+
+                                if (requests.Count > _configuration.QueryEmbeddingsMaxBatchSize)
```

Comment:
Here you _already_ passed the max batch size, so that isn't a good place to be in.
Also, what if we have a _single_ request that has more than the max size?